File: python/0093复原IP地址.py


def restoreIpAddresses(s: str) -> list:
    res=[]
    temp=[]

    def _isValid(s:str):
        if(any([s[0]=='0' and len(s)>1,int(s)>255])):
            return False
        return True

    def _backTracking(start:int):
        nonlocal s
        
        if(len(temp)==4 and sum([len(i) for i in temp])==len(s)):
            res.append(temp.copy())
            return

        for i in range(start,len(s)):
            frame=s[start:i+1]
            if(_isValid(frame)):
                temp.append(frame)
                _backTracking(i+1)
                temp.pop()

    _backTracking(0)
    return [".".join(i) for i in res]
    

print(restoreIpAddresses("25525511135"))

File: cpp/0093复原IP地址.cpp
#include<vector>
#include<string>
#include<iostream>
using namespace std;

vector<string> temp;
vector<string> res;

bool IsValid(string& s)
{
    if (s[0] == '0' and s.size() > 1)
        return false;
    if(s.size()<1||s.size()>3)
        return false;
    if (stoi(s) > 255)
        return false;
    return true;
}


void BackTracking(string &s,int start=0)
{
    if(temp.size()==4)
    {
        string ss;
        for(string s:temp)
        {
            
            ss.append(s + ".");
        }
        if(ss.size()==s.size()+4)
        {
            
            ss.pop_back();
            res.push_back(ss);
            return;
        }
    }

    for (int i = start;i<s.size();++i)
    {
        string ss = s.substr(start, i - start + 1);
        if (IsValid(ss))
        {
            // cout << 1;
            temp.push_back(ss);
            BackTracking(s,i + 1);
            temp.pop_back();
        }
    }
}

vector<string> restoreIpAddresses(string s)
{

    BackTracking(s);
    return res;
}

int main()
{
    restoreIpAddresses("25525511135");

    for (string s :res)
    {
        cout << s;
    }
        
}